lib/elixir/lib/macro.ex

```diff
@@ -2280,4 +2280,138 @@ defmodule Macro do
   defp trim_leading_while_valid_identifier(other) do
     other
   end
+
+  @doc """
+  Default backend for `Kernel.dbg/2`.
+
+  This function provides a default backend for `Kernel.dbg/2`. See the
+  `Kernel.dbg/2` documentation for more information.
+
+  This function:
+
+    * prints information about the given `env`
+    * prints information about `code` and its returned value (using `opts` to inspect terms)
+    * returns the value returned by evaluating `code`
+
+  You can call this function directly to build `Kernel.dbg/2` backends that fall back
+  to this function.
+  """
+  @doc since: "1.14.0"
+  @spec dbg(t, t, Macro.Env.t()) :: t
+  def dbg(code, options, %Macro.Env{} = env) do
+    header = dbg_format_header(env)
+
+    quote do
+      to_debug = unquote(dbg_ast_to_debuggable(code))
+      unquote(__MODULE__).__dbg__(unquote(header), to_debug, unquote(options))
+    end
+  end
+
+  # Pipelines.
+  defp dbg_ast_to_debuggable({:|>, _meta, _args} = pipe_ast) do
+    value_var = Macro.unique_var(:value, __MODULE__)
+    values_acc_var = Macro.unique_var(:values, __MODULE__)
+
+    # Using the private version of unpipe/2 here because Macro.unpipe/1 is deprecated.
+    [start_ast | rest_asts] =
+      asts = for {ast, 0} <- :lists.reverse(unpipe(pipe_ast, _acc = [])), do: ast
```

Comment:
Ironically I deprecate Macro.unpipe because it would eagerly expand and break dbg... but now dbg is the one using it. :D So I have undeprecated it on main.

```suggestion
    [start_ast | rest_asts] = asts = for {ast, 0} <- unpipe(pipe_ast), do: ast
```